packages/go_router/lib/src/delegate.dart

```diff
@@ -278,8 +279,25 @@ class ImperativeRouteMatch extends RouteMatch {
     required super.error,
     required super.pageKey,
     required this.matches,
-  });
+  }) : _completer = Completer<T?>();
 
   /// The matches that produces this route match.
   final RouteMatchList matches;
+
+  /// The completer for the promise returned by [GoRouter.push].
+  final Completer<T?> _completer;
+
+  /// Completes the promise returned by [GoRouter.push].
+  @override
+  void complete([dynamic value]) {
+    _completer.complete(value as T?);
+  }
+
+  /// Returns `true` if the promise returned by [GoRouter.push] has been completed.
+  @override
+  bool didComplete() => _completer.isCompleted;
+
+  /// The future of the [RouteMatch] completer. When the future completes, this
+  /// will return the value passed to [complete].
+  Future<T?> get future => _completer.future;
```

Comment:
this can be private

Reply:
The problem is we need to return this completer future in the push function, if we make it private we cannot return it.

Reply:
You can access private method as long as the class is in the same file, so it should be fine